### PR TITLE
Add export transform event

### PR DESCRIPTION
**Pull request by licitdev**
## Description

Add a new filter event to allow for validation and manipulation during export transformation.
Used `multipart/form-data` to overcome the `MAX_PAYLOAD_SIZE`.
Removed `json2csv` from app dependencies.

image

## Type of Change

- [ ] Bugfix
- [x] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [x] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [x] Documentation was added/updated -> https://github.com/directus/docs/pull/125

### CODE REVIEW

**Review by br41nslug — DISMISSED** [review]
When testing this event using a basic flow im getting the full flow context exported to my csv. The same happens for the flow provided in the PR description.
image
with the script being:
```
module.exports = async function(data) {
	return { "payload": [{ "intercepted": "because i can now" }] };
}
```
Results in the following export:
```
"$trigger.payload","$trigger.event","$trigger.format","$trigger.options","$last.payload","$accountability.user","$accountability.role","$accountability.admin","$accountability.app","$accountability.ip","$accountability.userAgent","$accountability.origin","$accountability.share","$accountability.share_scope","$accountability.permissions","log_k4w35","exec_iqzp6.payload"
"[{""id"":1,""img"":""4421ebeb-6970-4a65-bed0-353e040e4c09"",""file"":""387037c6-91f3-41c7-aad3-c89e43bd1f09"",""toggle"":true,""files"":[1,2,5,6,7]}]","export.transform","csv",,"[{""intercepted"":""because i can now""}]","5d336e28-d245-4252-8456-1d1e9799f05d","dbe96f38-b0a8-4bdd-8d10-5d2a3a52a37d",true,true,"127.0.0.1","Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:105.0) Gecko/20100101 Firefox/105.0",,,,"[]",,"[{""intercepted"":""because i can now""}]"
```

**Comment by licitdev** [comment]
@br41nslug Possible to double check on your trigger's response body if it's set to `Data of Last Operation`?

image

Tested your flow and gotten the following export.

image

**Comment by br41nslug** [comment]
You're very right i had it on "All Data" 😅

**Review by azrikahar — COMMENTED** [review]
LGTM! Tested a a simple Flows setup:

image

with the following non-exhaustive-demo-only script:

chrome_2sUbKHqP7K

and the exports are sanitized as expected 🙌

| Before (CSV) | After (CSV) |
|---|---|
| chrome_3N7ecchPCw | chrome_1vzgkryxbQ |

File: api/src/controllers/utils.ts [hunk 1]
```diff
@@ -163,6 +163,56 @@ router.post(
...
+
+		busboy.on('error', (err: Error) => next(err));
+
+		busboy.on('close', async () => {
+			if (!inputString) {
+				return next(new ForbiddenException());
+			}
+
+			try {
+				const input = JSON.parse(inputString);
```
[the remark this answers is not in the document]
Comment (licitdev) [reply api/src/controllers/utils.ts]: Refactored. Thanks!

**Review by br41nslug — DISMISSED** [review]

**Comment by rijkvanzanten** [comment]
Other than that, this LGTM 🙂

File: app/src/utils/save-as-csv.ts [hunk 2]
```diff
@@ -63,20 +64,32 @@ export async function saveAsCSV(collection: string, fields: string[], items: Ite
...
 
-	const dateString = `${now.getFullYear()}${(now.getMonth() + 1).toString().padStart(2, '0')}${now
-		.getDate()
-		.toString()
-		.padStart(2, '0')}`;
+		const response = await api.post('/utils/transform-export?export=csv', formData, {
+			headers: {
+				'Content-Type': 'multipart/form-data',
+			},
+		});
```
[the remark this answers is not in the document]
Comment (licitdev) [reply app/src/utils/save-as-csv.ts:82]: Used `multipart/form-data` to overcome the `MAX_PAYLOAD_SIZE`.

[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply app/src/utils/save-as-csv.ts:82]: Ah interesting. Did you run into the default max payload size often? So far we've only used multipart/form-data for binary data (eg files), but never for JSON IO (as that's always application/json), so I'm wondering if we instead want to raise the default max payload size a bit and just use json as alway s

[the remark this answers is not in the document]
Comment (licitdev) [reply app/src/utils/save-as-csv.ts:82]: Yes, the `MAX_PAYLOAD_SIZE` was faced also with schema endpoints. The limit is heavily dependent on the data to be exported which might contains `text` type fields with a larger size than expected. Going the multipart way eliminates allowing too big a `MAX_PAYLOAD_SIZE` for requests to other services. 🤔

[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply app/src/utils/save-as-csv.ts:82]: > Going the multipart way eliminates allowing too big a MAX_PAYLOAD_SIZE for requests to other services. 🤔

Yeah, but it also defeats the point of having MAX_PAYLOAD_SIZE in the first place, as you're still parsing the incoming payload as JSON anyways 🤔

[the remark this answers is not in the document]
Comment (licitdev) [reply app/src/utils/save-as-csv.ts:82]: Yes I agree. The same issue was faced with schema endpoints and to overcome it was to do a file upload instead. With the schema endpoints switching to use diffs instead, the `MAX_PAYLOAD_SIZE` limit could be easily hit. That's if JSON is the selected output. If YAML is the selected output, it utilises the multipart upload too.

Possible "solutions"
1. switch to use `application/json` and state in docs about payload size errors 
2. switch endpoint to use proper file upload for no limits (multipart fields has default 1mb limit)

[the remark this answers is not in the document]
Comment (licitdev) [reply app/src/utils/save-as-csv.ts]: Switched back to using `application/json` as `MAX_PAYLOAD_SIZE` is increased to `1mb` in https://github.com/directus/directus/pull/16039

File: api/src/controllers/utils.ts [hunk 3]
```diff
@@ -163,6 +163,24 @@ router.post(
...
 
+router.post(
+	'/transform-export',
+	asyncHandler(async (req, res, next) => {
+		if (req.is('application/json') === false || !req.headers['content-type']) {
+			throw new InvalidPayloadException(`Payload should be a JSON object`);
+		}
+
+		if (!req.sanitizedQuery.export) {
+			throw new ForbiddenException();
```
Comment (br41nslug) [inline api/src/controllers/utils.ts:174]: I think Forbidden may not be the most appropriate error here, `InvalidQuery` perhaps

**Review by br41nslug — COMMENTED** [review]
besides the error thrown looking good to me 😄

File: api/src/controllers/utils.ts (hunk 3 above)
Comment (licitdev) [reply api/src/controllers/utils.ts]: Refactored, `InvalidQuery` should be it. Thanks! 🙂

File: api/src/controllers/utils.ts [hunk 4]
```diff
@@ -163,6 +163,24 @@ router.post(
...
+		if (!req.sanitizedQuery.export) {
+			throw new InvalidQueryException('"export" has to be provided.');
+		}
+
+		res.locals.payload = { data: req.body };
+
+		return next();
+	}),
+	respond
+);
```
Comment (rijkvanzanten) [inline api/src/controllers/utils.ts:182]: This whole endpoint is a no-op now, as the transform happens based on the ?export flag. I get why that's the case, but it feels a little.. odd. 

Also should we maybe call it `/utils/transform` (drop the `-export`) and allow the user to pass a `?format` flag similar to `?export` 🤔 

I'm probably overthinking this 🤷🏻

**Comment by rijkvanzanten** [comment]
@licitdev Should we pick this back up, or close it out for now and circle back later?

**Comment by licitdev** [comment]
Closing stale PR. A different approach should be explored to prevent the cross-site scripting vulnerability.

**Pull request closed without merging**